content/languages/cpp/authoring/index.md

```diff
@@ -0,0 +1,387 @@
+
+This article is not a standalone tutorial on creating kata or translations. It's meant to be a complementary, C++-specific part of a more general set of HOWTOs and guidelines related to [content authoring](/authoring/). If you are going to create a C++ translation, or a new C++ kata from scratch, please make yourself familiar with the aforementioned documents related to authoring in general first. 
+
+## General info
+
+Any technical information related to the C++ setup on Codewars can be found on the [C++ reference](/languages/cpp/) page (language versions, available libraries, and setup of the code runner).
+
+
+## Description
+
+C++-specific code can be inserted with the `cpp` flag in [sequential code blocks](/references/markdown/extensions/#sequential-code-blocks):
+
+~~~
+```cpp
+
+...your code here...
+
+```
+~~~
+
+C++-specific paragraphs can be inserted with [language conditional rendering](/references/markdown/extensions/#conditional-rendering):
+
+```
+~~~if:cpp
+
+...text visible only for C++ description...
+
+~~~
+
+~~~if-not:cpp
+
+...text not visible in C++ description...
+
+~~~
+```
+
+## Tasks and Requirements
+
+Some concepts don't always translate well to or from C++. C++ allows for a variety of paradigms and techniques, many of which are not necessarily suited to other languages. For this reason, certain constructs should be avoided and some kata just should not be translated to C++.
+- C++ is statically typed, so any task that depends on dynamic typing can be difficult to translate into C++, and attempts to force a C++ kata to reflect a dynamically typed interface can lead to ideas that enforce a really bad design.
+- There are only a few additional libraries available to the C++ runner, so kata that take advantage of specialized packages installed for other languages become much more difficult to implement in C++.
+- Many features of C++ do not have direct equivalents in other popular languages: mixture of allowed paradigms, template meta-programming, unmanaged memory backed by RAII, native access to the platform and runtime, and many others. C++ kata which rely on them can be difficult to translate to other languages.
+
+
+## Kata snippets and template
+
+Due to the complicated compilation model of C++ code, paired with the fact that Codewars strictly enforces the names, amount, and meaning of source files, the code runner uses a [template](/languages/cpp/solution_template) to concatenate all kata snippets into a single translation unit. This fact has various consequences on the code, some positive, and some negative: namespace pollution, symbols introduced by one snippet being visible in other snippets, etc. However, in the majority of cases, it does not affect kata in any way, and whenever possible, this behavior should be treated as an implementation detail of the C++ code runner. Kata snippets should be treated as separate source files, if possible.
```

Comment:
I don't know why it was decided to concatenate at first, but I don't think there's any reason to do this (other than we have 1K+ kata that relies on the setup). Maybe I'm missing something.

Reply:
I can see a few reasons for this, but I believe the most probable one is handling of templates in C++. Templates have to be defined in every translation unit, and that's why they're always located in header files. If a solution would need to define a template, it would be not enough for a solution to be a cpp file (i.e. a separate translation unit).

Reply:
Ah, I see. If it was possible to have `solution.h` and `solution.cpp`, the test can include `solution.h` to be compiled separately and linked. Since we can't do that, we don't have a choice.

Reply:
I don't think we need a separate page for the template snippet. There's nothing much to explain further, and the new reusable one is self-explanatory.

Reply:
Solution snippet is not the only one affected by this issue. It applies also to, for example, preloaded, which might want to define global/static data in a cpp part, and declare it in a header file.

> Ah, I see. If it was possible to have `solution.h` and `solution.cpp`, the test can include `solution.h` to be compiled separately and linked. Since we can't do that, we don't have a choice.

It _could_ be done, but in kinda awkward way: snippets could contain some kind of markers which would separate "header" part and ".cpp part", and runner would split them into files appropriately. It's clunky, awkward, but in 99% not necessary, and some default behavior would suffice (no markers and store everything as a cpp file, compile, and link). Only those who need it, could use something like:

```cpp
/*============ solution.h ============*/
// ... declarations, template definitions

/*===========solution.cpp============*/
// ... implementation part
```

Not that I like this solution, but it _is_ one of solutions.



> I don't think we need a separate page for the template snippet. There's nothing much to explain further, and the new reusable one is self-explanatory.

Ok so I will remove the link to the "solution template" page for now, and later I will turn the "solution template" page into "how to train on cpp kata locally" and add the link back.